internal/provider/resource_tfe_workspace_execution_mode.go

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func resourceTFEWorkspaceExecutionMode() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTFEWorkspaceExecutionModeCreate,
+		Update: resourceTFEWorkspaceExecutionModeUpdate,
+		Read:   resourceTFEWorkspaceExecutionModeRead,
+		Delete: resourceTFEWorkspaceExecutionModeDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"agent_pool_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"workspace_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"execution_mode": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						"agent",
+						"local",
+						"remote",
+					},
+					false,
+				),
+			},
+		},
+	}
+}
+
+func resourceTFEWorkspaceExecutionModeCreate(d *schema.ResourceData, meta interface{}) error {
```

Comment:
I feel like either resourceTFEWorkspaceExecutionModeCreate should call resourceTFEWorkspaceExecutionModeUpdate. They both perform the exact same operation.

Reply:
I wondered if actually the `Update` should use the d.HasChange() on `agent_pool_id` and if there's been no change on the pool ID then throw an error? Otherwise, stuff the new pool ID into the options struct? @brandonc

Reply:
I think it's OK to send everything, just like on create. I think you'll want to conditionally send the agent pool ID, though, to only send it if you have one. Whatever makes the API and state file happy.